DLQ Service (GSI-1073) (#128)
ghga-de/epic-docs: 61-matamata-turtle/technical_specification.md

### Not included:
- A web-based user interface

## API Definitions:

### RESTful/Synchronous:

When interacting with DLQ topics via the API, only the service abbreviation and the 
plain topic name are required.

- `GET /{service}/{topic}`
  - *Returns the next events in the topic.*
    *This is a preview of the events and does not impact event ordering within the DLQ.*
  - Auth Header: internal token
  - Query parameters:
    - `skip` *(int)*: pagination param to skip the first `skip` results
    - `limit` *(int)*: the maximum number of events to preview per 'page'
  - Response Body: array of event objects
  - Response Status:
    - `200 OK`: The preview was successful
    - `401 Unauthorized`: auth error (not authenticated)
- `POST /{service}/{topic}`
  - *Processes the next event in the topic, optionally publishing the supplied event.*
    *Returns the payload passed to the publisher.*
  - Auth Header: internal token
  - Query parameters:
    - `dry_run` *(bool)*: if True, the endpoint will not actually publish the event to
    the retry topic or delete it from the database. This is useful for verifying what
    will get published before resolving an event.

>   - Request Body: empty OR JSON representation of corrected event

Author: That's correct

  - Response Body: the event that was/would be published.
  - Response Status:
    - `200 OK`: The event test was successful
    - `422 Unprocessable Entity`: The non-empty request body is not a valid event. 
    - `401 Unauthorized`: auth error (not authenticated)
- `DELETE /{service}/{topic}`
  - *Directly discards the next event in the topic*
  - Auth Header: internal token
  - Response Body: empty
  - Response Status:
    - `204 No Content`: The event has been discarded
    - `401 Unauthorized`: auth error (not authenticated)

> 

Author: That's fair. Maybe we should replace the composite event ID with a UUID4? Then we could generate that UUID upon publish, similar to the CID, and debug-log it for fuller traceability (that'd be a hexkit change).
Given that discarding an event means we're simply done with it forever, should we be allowed to discard events even if they're not the "next" event for a topic?

> 

Author: Another thought: Since changing the event_id would technically be a breaking change, would it be acceptable to provide the 4 components as query or path parameters? 
E.g. `DELETE /{service}/{topic}?partition=0&offset=123` or  `DELETE /{service}/{topic}/0/123`?

> 

Reviewer: Yes, that would work as well.

> All endpoints should require an internal auth token.

Author: Data stewards should never touch the DLQ. It should only be people from the architecture team. But I think our data stewards have no interest in interacting with the DLQ anyway, so modifying the auth token, commons, and auth service is not something we need to do on first implementation IMO

  the DLQ topic. This is crucial functionality that enables the user to decide how to
  resolve the next event.

### DLQ Sequence Illustrated
![DLQ Flow](./images/db%20dlq%20flow.png)

1. A service tries to consume a newly published event, but encounters an error.
2. The service tries again, also unsuccessfully.
3. Retries are exhausted, so the service publishes the event to the global DLQ topic.
4. The DLQ Service consumes the event from Kafka and stores it in the database.
   1. This involves extracting/adding the 'service' and 'event_id' fields.
5. Developers query the DLQ Service via the HTTP API to resolve an event.
6. The DLQ Service gets the event from the DB and does basic validation.
7. The event is sent to the original service's retry topic.
8. The DLQ Service returns the published event information to the client.
9. The service consumes the previously-DLQ'd event from its retry topic.

### DLQ Topic Arrangement
There is one DLQ topic. The DLQ topic name is set in configuration for some degree of
flexibility should it be needed, but all services should use the same configured value.
The default is `dlq`.

### Persisting DLQ Events

> The DLQ Service will continually listen to the DLQ topic. When it gets an event,

Reviewer: ```suggestion
The DLQ Service will continually consume from the DLQ topic. When it gets an event,
```

it will immediately store the event in the database.
Under the initial implementation, all events will go into a single collection.
MongoDB's `aggregate` functionality will be used (as in `mass`) to pull back the
correct events for a given `service` and `topic`, sorted by timestamp.

### Event Ordering:
Dead letter queues inherently present a potential threat to system-wide event ordering.
However, ordering events by keys, the idempotent design of our services, and
sorting events by timestamp (oldest first) prevents sequence problems
as long as events are designed to use the correct keys and topics in the first place.

### Event identification:
Right now, events can be identified through a combination of correlation ID + event
type and topic. It would be easier if there were a single field to associate a given
event in a DLQ topic with its source event in the original topic. The combination of
service name, topic name, partition, and offset uniquely identify an event, so we
can concatenate the values in a single string and pass it along as an event header.

> E.g.: `dcs,file-registrations,0,123` would indicate an event from the "file-registrations"

Reviewer: I wonder whether we should use a different separator that does not need to be encoded when passed in a URL, so the URL is more readable. We cannot use dot or hyphen since they can occur in topic names. Underscores should be possible according to our naming conventions, but we should not rely on that. Maybe a tilde `~`?

Author: Like I said in my response to your comment on the DELETE endpoint, maybe this is enough pressure to simply switch to a UUID.

Reviewer: OTOH the composite event id was more readable and meaningful when debugging. But maybe we can provide the same information in other fields, similar to the exception?

Author: Per your idea below, why not leave that field like it is and include a `dlq_id` in previewed events?

Reviewer: If we don't use it as an id, then I would prefer separate fields for readability and parseability.

Author: For now, the event ID will stay like it is, but the distinct elements within it will be stored in separate fields when the DLQS stores the DLQ event in the DB.

topic, partition 0, offset 123. The information is only seen by the consumer, not by the
publisher, which is the only downside of doing it this way instead of generating a UUID
right before publishing the event.

The event ID is always created, but only propagated if it enters the DLQ cycle.
This differs from the correlation ID, which is propagated across services and persisted
in the database in the case of outbox events.

> To illustrate:

Reviewer: Mention here that "nos" is our example service and "users" the example topic, where another services publishes and event to users, and nos tries to consume it.

Author: Done

> ![Event ID usage](./images/event%20ID.png)

Reviewer: Would be good to have little numbers in the diagram like in the other one.

Author: Done

1. An event is published to the "users" topic, where it is stored at partition 0, offset 17.
2. The `hexkit` Kafka provider used by the `NOS` gets the message from `aiokafka`.
3. The Kafka provider hands the transformed event over to the service-defined
   translator, which performs some business logic driven by the event.
4. The translator encounters an error which it cannot mitigate, and all retries fail.
5. The Kafka provider logs the error, including the event ID (`"nos,users,0,17"`).

> 6. The Kafka provider publishes the event with the event ID header to the DLQ topic.

Reviewer: Why is the DLQ in the diagram labelled "users.nos-dlq"? I thought it was global and just `dlq`?

Author: I just missed it

8. We learn the problem was actually a database issue that we've since fixed.
9.  We use the DLQ service to publish the event to the retry topic, sans event ID header.
   Instead, the DLQ service includes a special header with the original topic ("users").
10. The `NOS` encounters the republished event, this time from its dedicated retry
  topic. The Kafka provider understands that the retry topic is special, so it obtains
  the topic field from the original topic header. If the retry event fails again, the
  old event ID would be misleading had we modified the event in the DLQ service.

If the translator encounters no errors, the event ID is not needed and never created.

### Event processing:
When we resolve the next event in a given DLQ topic, it can go one of two ways:
- We can discard the event, meaning it is not republished to a retry queue and is
  effectively ignored.
- We can process the event, where the event is published to the retry topic or
  discarded (e.g. if the service info is missing). The processing can also be bypassed
  by directly supplying a JSON event to the API, which publishes the event outright
  as long as the correlation ID matches.

> Another useful piece of information for the DLQ service is the exception information.

Reviewer: In principle, the exception information should be available in the log of the service and we should be able to find it using the timestamp and event ID. So this would be just for convenience, so that it can be shown in the DLQ service client?

Author: Yeah, it doesn't cost anything and I think it would be nice to have

Reviewer: Could we provide the information about the service name, topic, partition and offset in a similar way if we use a random UUID to identify the event?

Author: Sure. They could be separate fields:
```
{
    "topic": "dlq",
    "type_": "file_registered",
    "payload": {},
    "key": "",
    "timestamp": 123412341234,
    "headers": {
        "correlation_id": "f7b5b22b-ed2a-4603-9466-2e3ae4178e42",
        "event_id": "2be69da8-84c1-453f-9224-10f7fdeb9328" (or "dcs,file-registrations,0,123"),
        "service": "dcs",
        "original_topic": "file-registrations",
        "partition":0,
        "offset": 123,
        "exc_cls": "RuntimeError",
        "exc_msg": "useful error message"
    }
}
```

Alternatively, all those "helper" datum could go into a subdict called something like `_dev_info`, where the underscore helps to...underscore... that it's not meant to be used programmatically (just an idea, maybe it doesn't make sense).
```
{
    "topic": "dlq",
    "type_": "file_registered",
    "payload": {},
    "key": "",
    "timestamp": 123412341234,
    "headers": {
        "correlation_id": "f7b5b22b-ed2a-4603-9466-2e3ae4178e42",
        "event_id": "2be69da8-84c1-453f-9224-10f7fdeb9328" (or "dcs,file-registrations,0,123"),
        "service": "dcs",
        "original_topic": "file-registrations",
        "_dev_info": {
            "partition":0,
            "offset": 123,
            "exc_cls": "RuntimeError",
            "exc_msg": "useful error message"
        }
    }
}
```

Reviewer: Or maybe put `dlq_info` on the top level (flat is better than nested)?

Reviewer: And another field `dlq_id` that would contain the UUID for the event?

Author: The revised approach inside the DLQS will put all the dlq-specific info (partition, offset, exc info, etc.) into a top-level `dlq_info` field, and generate a DLQ-specific UUID for reference to put in a top-level `dlq_id` field.

There is currently no mechanism to provide that to the DLQ service, but the info could
be passed through event headers. The exception header would then be removed upon
republishing the event to a retry topic.

This [Java DLQ implementation](https://medium.com/nerd-for-tech/-to-re-queue-apache-kafka-dlq-messages-95941525ca77)
uses such headers (toward the bottom).

### Processing Dry Run:
It would be nice to be able to perform a dry-run of event processing as a sanity check.
The test endpoint would perform any applicable processing on the next event and
return the result instead of publishing it. If processing would result in the

> event being discarded instead of published, then it will return nothing.  

Reviewer: Above you already mention this as a feature to be implemented instead of a nice to have. Maybe just remove this section.

### Previewing Events:
Previewed events will be formatted with the `ExtractedEventInfo` class and returned as 
JSON:
```json
{

>   "topic": ...,

Reviewer: could also have a unique id (the one used in the database anyway) that can be passed to the POST and DELETE endpoints instead of the event ID

Author: This is also a possibility, and would keep the change within the DLQS

  "type_": ...,
  "payload": {
    "field1": ...,
    "field2": ...,
    ...
  },
  "key": ...,
  "headers": {
    "correlation_id": ...,
    "exc_class": ...,
    "exc_msg": ...,

>     "id": <topic - partition - offset>

Reviewer: the id would also contain the service name, right? would be good to show the exact format of the id here again.

    - `401 Unauthorized`: auth error (not authenticated)

Each endpoint will return a `404` status code if the `service` parameter does not match
an existing service, or the `topic` parameter is not configured for the given service.

All endpoints should require an internal auth token.

### Payload Schemas for Events:

**Outbound from DLQ Service**:  
The outbound events published to retry topics will look exactly like other Kafka events
published or consumed by the GHGA microservices, except they will feature the
`original_topic` header.

**Inbound to DLQ Service**:  
The inbound events in the DLQ topics will feature additional
headers to communicate exception information like the message and class name. That will
require a small modification to `hexkit`.

**Previewed Events**:  
Events returned by the preview endpoint will match the format defined by the
[`ExtractedEventInfo`](https://github.com/ghga-de/hexkit/blob/e3cad2d57212d1ba897e620501d5d2b120c70338/src/hexkit/providers/akafka/provider/eventsub.py#L63) dataclass.

**Other**:  
The event ID header (consisting of the original topic, partition, and offset) will be
featured in both inbound and outbound events.

> 

Author: Events published to the DLQ topic will have the original topic header

- *Requeue/Republish an event*: In the context of the DLQ service, this means to publish
  the next Kafka event from a given DLQ topic to the corresponding retry topic.
- *Process an event*: Consume the event and apply validation and processing logic (if
  applicable). The event will either be republished or discarded as a result.
- *Discard an event*: To decline to requeue/republish an event to the retry topic. This
  can occur upfront via a direct command by the user, or it can occur as a result of
  the event-processing logic.
- *Resolve an event*: Process or discard an event.
- *Preview events*: Peek at the next N events without affecting the order of events in
  the DLQ topic. This is crucial functionality that enables the user to decide how to
  resolve the next event. The number of events previewed is capped by configuration.

### DLQ Topic Arrangement

For a given Kafka topic, there will be a separate DLQ topic *per service*.
For example, the UCS, DCS, and IFRS subscribe to the topic containing File Deletion
Requested events. They would publish failed File Deletion Requested events to their own
DLQ topics, e.g. `file-deletions.ucs-dlq`, `file-deletions.dcs-dlq`,

> `file-deletions.ifrs-dlq`. Each service has only one retry topic, however:

Author: Reworded for clarity -- let me know if you think it needs further work

> `file-deletions.ifrs-dlq`. Each service has only one retry topic, however:

Author: This is now outdated as there is only one global dlq topic. One retry topic per service is the same though.

![DLQ Topic Arrangement](./images/topic%20distribution.png)

### Event Ordering:
Dead letter queues inherently present a potential threat to system-wide event ordering.
However, ordering events by keys, the idempotent design of our services, and having

> separate DLQ topics for each original topic *per service* should prevent most problems.

Author: Just unknown unknowns or flawed event design on our part, otherwise no.

Additionally, only one partition should be configured per DLQ topic as there will be

> only one consumer for each DLQ topic.

Author: Yes it is. I reworded this part as a requirement to sound less like a suggestion

- `hexkit`: Add exception information to DLQ event headers
- `hexkit`: Add preview functionality to `KafkaDLQSubscriber`
- `hexkit`: Add preview dry-run functionality to `KafkaDLQSubscriber`

### Optional:
- Add event ID header in `hexkit`, consisting of the original topic, partition, and
offset

### Not included:
- A web-based user interface

## API Definitions:

### RESTful/Synchronous:

DLQ topics consist of the original topic name, the service abbreviation, and the fixed
suffix `-dlq`. The latter could technically be omitted, but using it visually
distinguishes dlq topics from others and making debugging easier.
When interacting with DLQ topics via the API, however, the above structure does not need
to be remembered: only the service abbreviation and the plain topic name.

- `GET /services`
  - *Returns a list of all configured services*
  - Auth Header: internal token
  - Response Body: array of strings (service names)
  - Response Status: `200 OK`
- `GET /topics/{service}`
  - *Returns a list of all topics configured for the chosen service*
  - Auth Header: internal token
  - Response Body: array of strings (topic names)
  - Response Status: `200 OK`

> - `GET /{service}/{topic}`

Reviewer: Should we add query params here to allow pagination (limit and offset)?

Author: I've made the change

    *This is a preview of the events and does not impact event ordering within the DLQ.*
    *Like the test endpoint below, this endpoint is idempotent.*
  - Auth Header: internal token
  - Response Body: array of event objects
  - Response Status:
    - `200 OK`: The preview was successful
    - `401 Unauthorized`: auth error (not authenticated)
- `GET /test/{service}/{topic}`
  - *Returns the result of processing the next event in the specified topic. Nothing*
    *is published and offsets are not affected. Calling the same test path multiple*
    *times will return the same result.*
  - Auth Header: internal token
  - Response Body: the event that would be published, empty if event would be discarded.
  - Response Status:
    - `200 OK`: The event test was successful
    - `422 Unprocessable Entity`: The non-empty request body is not a valid event. 
    - `401 Unauthorized`: auth error (not authenticated)
- `POST /{service}/{topic}`
  - *Processes the next event in the topic, optionally publishing the supplied event*
  - Auth Header: internal token
  - Request Body: empty
  - Response Body: empty or JSON representation of corrected event

>     - Need to consider whether checks should be added for fields like correlation ID.

Reviewer: Shouldn't this go to the *request* body?

Author: Yes, that's a typo
